jms/org.wso2.carbon.transport.jms/src/main/java/org/wso2/carbon/transport/jms/factory/JMSClientConnectionFactory.java

```diff
@@ -0,0 +1,226 @@
+            }
+        }
+
+        if (clientCaching) {
+            connections = new ArrayList<>();
+            initSessionPool();
+        }
+    }
+
+    public int getMaxNumberOfConnections() {
+        return maxNumberOfConnections;
+    }
+
+    public int getMaxSessionsPerConnection() {
+        return maxSessionsPerConnection;
+    }
+
+    /**
+     * Initialize the session pool with provided configuration
+     */
+    private void initSessionPool() {
+        SessionPoolFactory sessionPoolFactory = new SessionPoolFactory(this);
+
+        //create pool configurations
+        GenericObjectPoolConfig config = new GenericObjectPoolConfig();
+        config.setMaxTotal(maxNumberOfConnections * maxSessionsPerConnection);
+        //todo: set the ideal limit and make the idle sessions timedout
+        config.setMaxIdle(maxNumberOfConnections * maxSessionsPerConnection);
+        config.setBlockWhenExhausted(true);
+        config.setMaxWaitMillis(poolWaitTimeout);
+
+        //initialize the pool
+        sessionPool = new GenericObjectPool<SessionWrapper>(sessionPoolFactory, config);
+    }
+
+    private void setCache(Properties properties) {
+        String cacheLevel = properties.getProperty(JMSConstants.PARAM_JMS_CACHING);
+        if (null != cacheLevel && !cacheLevel.isEmpty()) {
+            this.clientCaching = Boolean.parseBoolean(cacheLevel);
+        } else {
+            this.clientCaching = Boolean.TRUE;
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public synchronized void notifyError(JMSException ex) {
+        logger.error("Error occurred in JMS Client Connections. Re-initializing the resources. " + ex.getMessage());
+        closeJMSResources();
+        initSessionPool();
+    }
+
+    /**
+     * @return Connections list
+     */
+    public List<ConnectionWrapper> getConnections() {
+        return connections;
+    }
+
+    /**
+     * Borrow an object from the session pool
+     *
+     * @return SessionWrapper instance
+     * @throws Exception Exception when borrowing an object from the pool
+     */
+    public SessionWrapper getSessionWrapper() throws Exception {
+        return sessionPool.borrowObject();
+    }
+
+    /**
+     * Return an object to the Session pool
+     *
+     * @param sessionWrapper SessionWrapper instance
+     */
+    public void returnSessionWrapper(SessionWrapper sessionWrapper) {
+        sessionPool.returnObject(sessionWrapper);
+    }
+
+    /**
+     * Close cached JMS resources allocated for this Connection Factory
+     */
+    private void closeJMSResources() {
+        if (clientCaching) {
+            sessionPool.clear();
+            connections.forEach(connection -> {
+                try {
+                    connection.getConnection().close();
+                } catch (JMSException e) {
+                    logger.error("Error closing the connection" + e);
```

Comment:
Rather than logging `e` as a string shall we pass this as the throwable argument for `logger.error` method

Reply:
fixed